Aws test container

belosh59 opened the pull request:
Preliminary code which expose the current problems during test containers setup

0) DynamoDB discovery is working, you could try it.

1) At the moment we are creating dependencies for our selves (clients), so that it becomes difficult to inject localstack clients into the code. We need to expose private methods. But that also not working. Even deeper we also create such dependancy (S3 discovery is not working due to that perspective)
Otherwise with current clients we are trying to reach the real AWS. 

   1.1. I'm thinking about lightweight ServiceDiscovery pattern, but we need to keep clients being configured for different Regions being requested (e.g DynamoDB_AWS_GLOBAL / DynamoDB_US_WEST_1 ) 

2) Seems like AWS Java SDK doesn't provide possibility to override AWS Endpoint globally for all clients [link](https://github.com/aws/aws-cli/issues/4454)

3) For a demo purpose we could setup localstack separately but we need to bypass localstack docker URL to reach it. Meanwhile there no possibility for that in code

Proposal:
We could create our AWS Clients based on env variable being passed on the stage of execution - > if nothing passed fallback to the default one - https://aws.amazon.com/

[To not merge yet]

kickroot commented on magpie-aws/pom.xml:
```diff
@@ -207,6 +209,31 @@
     </dependency>
+    <dependency>
+      <groupId>software.aws.mcs</groupId>
+      <artifactId>aws-sigv4-auth-cassandra-java-driver-plugin</artifactId>
```
What's this needed for?

kickroot reviewed (commented):
For 1) I agree that you can use an env var to set the AWS endpoint.  Something like `MAGPIE_AWS_ENDPOINT`.  My first thought is to create a static utility method that will return the endpoint (defaulting to aws.amazon.com if not specified). Then each service will pass along that result to it's constructor.

For 3) Don't sweat the localstack demo env.  We'll cross that bridge when we get to it.

belosh59 commented on magpie-aws/pom.xml:
```diff
@@ -242,31 +269,35 @@
     <dependency>
-      <groupId>software.aws.mcs</groupId>
-      <artifactId>aws-sigv4-auth-cassandra-java-driver-plugin</artifactId>
-      <version>${aws-sigv4-auth-cassandra-java-driver-plugin.version}</version>
```
@kickroot  it is been already there, I have rearranged some dependencies to have better separation of purposes

PrzemyslawTusinski commented on magpie-aws/src/main/java/io/openraven/magpie/plugins/aws/discovery/services/S3Discovery.java:
```diff
@@ -42,11 +42,13 @@
 public class S3Discovery implements AWSDiscovery {
 
   private static final String SERVICE = "s3";
+  private final String RESOURCE_TYPE = "AWS::S3::Bucket";
```
Why did you move this to class member?

belosh59 replied:
Since such Strings is rather a discovery class related constant and used as is, I propose to not create it each time during invocation of the discovery. If will be agreed to do so that will be made in separate PR for all discoveries

PrzemyslawTusinski replied:
Defining RESOURCE_TYPE in function seems clearer to me.
For sure in approach proposed by you RESOURCE_TYPE have to be renamed to something else like RESOURCE_TYPE_S3_BUCKET. This is because there are cases with multiple RESOURCE_TYPEs in one class

belosh59 replied:
I will take off this change so far

PrzemyslawTusinski commented on magpie-aws/src/main/java/io/openraven/magpie/plugins/aws/discovery/AWSUtils.java:
```diff
@@ -200,7 +202,18 @@ public static GetMetricStatisticsResponse getCloudwatchMetricStatistics( String
     return StsClient.create().getCallerIdentity().account();
   }
+
+  public static <BuilderT extends AwsClientBuilder<BuilderT, ClientT>, ClientT> ClientT
+  apply(AwsClientBuilder<BuilderT, ClientT> builder, Region region) {
+    // Remap magpie clients to local environment
+    if (System.getProperty("aws.local.endpoint") != null) {
+      builder.endpointOverride(URI.create(System.getProperty("aws.local.endpoint")));
+    }
+    builder.region(region);
+    return builder.build();
+  }
```
This violates the rule of not having any test related code in running code.
Imo it's better to extract creating client before the given discover function like it's DynamoDbDiscovery. Then either production code or test code will create client needed for thier purpose without relying on checking env variable.

kickroot replied:
I don't think this is necessarily test related code so much as a way to allow users to change their AWS endpoint. It also happens to work for the test environment.  

Can you link to a code example @PrzemyslawTusinski of what you're thinking?

belosh59 replied:
@PrzemyslawTusinski  I was also thinking about kind of Dependency Inversion for discover function, to provide prepared clients targeting for different purpose (PROD vs TESTs), but:
1. We are creating clients depends on Regions for each iteration over configured Regions for discovery (AWSDiscoveryPlugin:98). So i doesn't looks good to me to create all region dependent clients before we start
2. Also we think about running Magpie against the Demo stack to not reach the real AWS for any kind of simulation.
But if you see better approach here I would rather

PrzemyslawTusinski replied:
Since we are using this code not only in test this approach is fine

kickroot replied:
Let's change the variable though, I don't like the `local` name and it also conflicts with how we assess other Magpie related variables.  It should be `MAGPIE_AWS_ENDPOINT`.

belosh59 replied:
But along with region we also configure endpointOverride property as well. Maybe `configure`. From that perspective `configure` name could be even more transparent about what is method about

kickroot replied:
I *think* region will get picked up automatically from the `AWS_REGION` environmental variable.

belosh59 replied:
@kickroot  there we are setting region for client from config.yml. For each chosen region where user requested discovery.

PrzemyslawTusinski commented on magpie-aws/src/main/java/io/openraven/magpie/plugins/aws/discovery/AWSUtils.java:
```diff
@@ -181,7 +183,7 @@ public static JsonNode add(List<? extends ToCopyableBuilder> responsesToAdd) {
   public static GetMetricStatisticsResponse getCloudwatchMetricStatistics( String regionID, String namespace, String metric, Statistic statistic, List<Dimension> dimensions) {
 
-    try (final CloudWatchClient client = CloudWatchClient.builder().region(Region.of(regionID)).build()) {
+    try (final CloudWatchClient client = apply(CloudWatchClient.builder(), Region.of(regionID))) {
```
Let's extract this CloudWatchClient creation before execution of discovery function. Like it's with given service client

belosh59 commented:
Added:
- CF update-stack command usage for dynamic resource configuration for various tests using readable `.yml` template format
- Fixed us-east-1 bucket bug processing
- Added PR review fixes based on suggestion

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/EC2DiscoveryIT.java:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  @Test
+  public void testEC2Discovery() {
```
whenRunningEc2DiscoveryShouldReturnInstanceEipSecurityGroupVolumeAndSnapshotInformation
It's probably good idea to split this test into 5. Each testing one thing

belosh59 replied:
Got your point @PrzemyslawTusinski . We rather have here a complete testing for chosen component from magpie discovery perspective, instead of the unit testing when we could write tests isolated for various corner cases.

Splitting works good but we are going to expose our private methods for component discovery of the various aspects.
For example how we made with IAM - but I dont really like to change private method to protected just for the test availability.

I threat those tests as e2e AWS component discovery based on localstack. It like an IT with WireMocks when you passing from the controller request to database layer and back verifying the response.
In any case I propose to change method naming convention here once we will extend them with additional coverage

PrzemyslawTusinski replied:
I didn't meant exposing private methods.
I meant running full discovery in each test case, then during asset phase assert only one resource on each test case

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/DynamoDbDiscoveryIT.java:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  @Test
+  public void testDynamoDBDiscovery() {
```
whenDiscoveringDynamoDbTablesShouldReturnProperContentInfo

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/disabled/KMSDiscoveryIT.java:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  @Test
+  public void testKMSDiscovery() {
```
Better naming?

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/disabled/CloudWatchDiscoveryIT.java:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  @Test
+  public void testCloudWatchDiscovery() {
```
Better naming?

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/IAMAccountDiscoveryIT.java:
```diff
@@ -0,0 +1,94 @@
+  private ArgumentCaptor<MagpieEnvelope> envelopeCapture;
+
+  @Test
+  public void testAccountDiscovery() {
```
Better naming?

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/IAMCredentialReportDiscoveryIT.java:
```diff
@@ -0,0 +1,62 @@
+  private ArgumentCaptor<MagpieEnvelope> envelopeCapture;
+
+  @Test
+  public void testCredentialReportDiscovery() {
```
Better naming?

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/IAMGroupDiscoveryIT.java:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  @Test
+  public void testGroupDiscovery() {
```
Better naming?

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/IAMPolicyDiscoveryIT.java:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  @Test
+  public void testPolicyDiscovery() {
```
Better naming?

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/IAMRoleDiscoveryIT.java:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  @Test
+  public void testRoleDiscovery() {
```
Better naming would be useful here

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/LambdaDiscoveryIT.java:
```diff
@@ -0,0 +1,89 @@
+   *  Unable to parse date format. Hence fallback to simpler template
+   */
+  @Test
+  public void testLambdaDiscovery() {
```
Better naming would be useful here

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/Route53DiscoveryIT.java:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  @Test
+  public void testRoute53Discovery() {
```
Better naming?

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/S3DiscoveryIT.java:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  @Test
+  public void testS3Discovery() {
```
Better naming?

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/SNSDiscoveryIT.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  @Test
+  public void testSNSDiscovery() {
```
Better naming?

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/SecretsManagerDiscoveryIT.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @Test
+  public void testSecretManagerDiscovery() {
```
Better naming?

PrzemyslawTusinski commented on magpie-aws/src/test/java/io/openraven/magpie/plugins/aws/discovery/services/SNSDiscoveryIT.java:
```diff
@@ -0,0 +1,72 @@
+    Mockito.verify(emitter, times(2)).emit(envelopeCapture.capture());
+    var invocations = envelopeCapture.getAllValues();
+    assertEquals(2, invocations.size());
+    assertTopic(invocations.get(0));
+    assertSubscription(invocations.get(1));
```
Please consider splitting this to 2 test cases

belosh59 replied:
This will be possible only if we will expose methods `discoverTopics` and `discoverSubscriptions`.
Not sure if we want it for test purpose. Initially I was designing cumulative approach to test input/output for available exposed method.
This decrease execution time and leading to the comparable coverage.
But it could fail because issue in either topic or subscription - at this point of IT testing that should be acceptable.
I will propose separate refactoring if required later.

belosh59 commented:
Builded and tested against the real AWS Services
